File: elmo_client.py
import time
import BaseHTTPServer
import httplib, urllib
import json

ELMO_IP 	= 'localhost'
ELMO_PORT 	= 3333
ELMO_ROUTES = []

class ELMO_HANDLER(BaseHTTPServer.BaseHTTPRequestHandler):
    def _set_headers(self):
        self.send_response(200)
        self.send_header('Content-type', 'text/html')
        self.end_headers()
 
    def do_GET(self):
		self._set_headers()		
		response = ""
		for ELMO_ROUTE in ELMO_ROUTES:
			if('/' + ELMO_ROUTE["route"] == self.path):
				response += ELMO_ROUTE["get"]()
		self.wfile.write(response)
 
    def do_HEAD(self):
        self._set_headers()
        
    def do_POST(self):
		self._set_headers()
		response = ""
		for ELMO_ROUTE in ELMO_ROUTES:
			if('/' + ELMO_ROUTE["route"] == self.path):
				response += ELMO_ROUTE["post"]()		
		self.wfile.write(response)
		
def add(route,type,get,post):
	ELMO_ROUTES.append({"route" : route, "type": type, "get" : get, "post" : post})

def register(ELMO_HOSTNAME,ELMO_HOSTPORT):
	data = []
	for route in ELMO_ROUTES:
		data.append({"route" : "http://%s:%s/%s" % (ELMO_IP, ELMO_PORT, route.route), "type" : route.type})
	headers = {"Content-type": "application/json","Accept": "text/plain"}
	conn = httplib.HTTPConnection(json.dumps(data))
	conn.request("POST", "", routes, headers)
	response = conn.getresponse()
	print response.status, response.reason
	conn.close()

def start():
    server_class = BaseHTTPServer.HTTPServer
    httpd = server_class((ELMO_IP, ELMO_PORT), ELMO_HANDLER)
    print time.asctime(), "Server Starts - %s:%s" % (ELMO_IP, ELMO_PORT)
    try:
        httpd.serve_forever()
    except KeyboardInterrupt:
        pass
    httpd.server_close()
    print time.asctime(), "Server Stops - %s:%s" % (ELMO_IP, ELMO_PORT)

File: elmo_client.js
/*
 * ELMO CLIENT MODULE  
 */
var http = require('http');
var querystring = require('querystring');

var ELMO_IP 	= 'localhost';
var ELMO_PORT 	= 3333;
var ELMO_ROUTES = [];

/*
 * Add an ELMO task
 */
exports.add = function(route,type,get,post) {

	ELMO_ROUTES.push({
			
		route : route,
		type : type,
		get : get,
		post : post
		
	});

}

/* 
 * Register ELMO tasks on ELMO server
 */
exports.register = function(ELMO_HOSTNAME, ELMO_HOSTIP) {
	
	var routes = [];
	for(i=0;i<ELMO_ROUTES.length;i++) {
		routes.push({
			route : 'http://'+ELMO_IP+':'+ELMO_PORT+'/'+ELMO_ROUTES[i].route, 
			type : ELMO_ROUTES[i].type
		});
	}
	var data = JSON.stringify(routes);
	var options = {
		host: 	ELMO_HOSTNAME,
		port: 	ELMO_HOSTIP,
		path: 	'/',
		method: 'POST',
		headers: {'Content-Type': 'application/json','Content-Length': Buffer.byteLength(data)}
	};
	var req = http.request(options, function(res) {
		console.log('Server responded');		
	});
	req.write(data);
	req.end();
	
}

/*
 * Start the ELMO client
 */
exports.start = function() {

	http.createServer(function (req, res) {
		
		var url = req.url; 
		
		if(req.method == 'POST') {
			
			var response = [];
			for(var i = 0; i < ELMO_ROUTES.length; i++ ) {				
				if('/' + ELMO_ROUTES[i].route == url) {
					response.push(ELMO_ROUTES[i].post(req));					
				}			
			}
			
			res.writeHead(200, {'Content-Type': 'text/html'});
			res.end(response.join());	
			
		} else {
			
			var response = [];
			for(var i = 0; i < ELMO_ROUTES.length; i++ ) {				
				if('/' + ELMO_ROUTES[i].route == url) {
					response.push(ELMO_ROUTES[i].get(req));					
				}			
			}
			
			res.writeHead(200, {'Content-Type': 'text/html'});
			res.end(response.join());
			
		}
		
	}).listen(ELMO_PORT);
	
}
